### PR TITLE
Add PROJECT_ENV

### PR DESCRIPTION
**Pull request by nevar**
See #587

### CODE REVIEW

File: core/erlc.mk [hunk 1]
```diff
@@ -60,14 +60,16 @@ app:: clean deps $(PROJECT).d
 endif
 
 ifeq ($(wildcard src/$(PROJECT_MOD).erl),)
+PROJECT_ENV ?= []
```
Comment (essen) [inline core/erlc.mk]: Define this here instead: https://github.com/nevar/erlang.mk/blob/38fe63dafdd8e9c902e5b242da4dd02ab0f0067f/core/core.mk#L27

File: test/core_app.mk [hunk 2]
```diff
@@ -28,6 +28,27 @@ core-app-appsrc-change: build clean
 	$t rm $(APP)/EXPECT
 endif
 
+ifndef LEGACY
+core-app-env: build clean
```
Comment (essen) [inline test/core_app.mk]: Alphabetical order.

File: test/core_app.mk [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # Core: Building applications.
 
-CORE_APP_CASES = appsrc-change asn1 auto-git-id erlc-exclude erlc-opts erlc-opts-filter error generate-erl generate-erl-include generate-erl-prepend hrl hrl-recursive makefile-change mib no-app no-makedep project-mod pt pt-erlc-opts xrl xrl-include yrl yrl-include
+CORE_APP_CASES = appsrc-change asn1 auto-git-id erlc-exclude erlc-opts erlc-opts-filter error generate-erl generate-erl-include generate-erl-prepend hrl hrl-recursive makefile-change mib no-app no-makedep project-mod pt pt-erlc-opts xrl xrl-include yrl yrl-include env
```
Comment (essen) [inline test/core_app.mk]: Alphabetical order.

**Comment by essen** [comment]
Small tidbits but otherwise looks fine. Also missing a section in the user guide.

Thanks!

**Comment by nevar** [comment]
updated

**Comment by nevar** [comment]
Faced with issue:
1. printf for app file don't escape % (will send PR for that)
2. For old make file ``$(shell cat file.env)`` eat all \n and if file has erlang comment then .app not right. ``$(file < file.env)`` look fine (don't eat \n). May by it is good to note that in document and add example how read file.env. I end up with ``$(shell cat file.env | sed 's/$$/\\n/')``

**Comment by nevar** [comment]
Also it will be good if we can autoadd ``file.env`` as dependency of ``ebin/$(PROJECT).app``

**Comment by essen** [comment]
What's the thing about % and " about? Not sure I get it.

**Comment by nevar** [comment]
If I have in file.env: ``[{test, "%"}]``, then printf will insert parameter for %. Same apply for ``PROJECT_DESCRIPTION = test %``.  Need make ``%`` -> ``%%``. This can fixed for all app_file template in https://github.com/ninenines/erlang.mk/blob/master/core/erlc.mk#L244. And this is separate from this PR I think.

If I have in file.env: ``[{test, "\"test\""}]``, then after insert to app template printf command will be: ``printf "[..skip..]{env, [{test, \"\\"test\\"\"}]}[..skip]" and after process we will get: ``{env, [{test, "\test\"}]``. Same apply for ``PROJECT_DESCRIPTION = test "\"test\""``. Need ``\`` -> ``\\``. Not sure we wan't fix that for PROJECT_DESCRIPTION.

**Comment by essen** [comment]
Right, thanks.

**Comment by nevar** [comment]
Change printf "..." to printf '...' for app_file.
Add escape / for PROJECT_ENV

**Comment by nevar** [comment]
#366

**Comment by nevar** [comment]
For get env from file and make version < 4: ``PROJECT_ENV = $(subst \n,$(newline),$(shell cat FILE.env | sed 's/$$/\\n/'))``

**Comment by essen** [comment]
CI currently running on this.

**Comment by essen** [comment]
Merged, thanks!!

**Pull request closed without merging**